lib/src/builder/parsing/declarations_from_members.dart

```diff
@@ -0,0 +1,385 @@
+// Copyright 2020 Workiva Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import 'ast_util.dart';
+import 'declarations.dart';
+import 'member_association.dart';
+import 'members.dart';
+import 'util.dart';
+import 'error_collection.dart';
+import 'version.dart';
+
+const overReactBoilerplateAnnotations = [
+  'Factory',
+  'Component',
+  'Component2',
+  'Props',
+  'State',
+  'AbstractComponent',
+  'AbstractComponent2',
+  'AbstractProps',
+  'AbstractState',
+  'PropsMixin',
+  'StateMixin',
+];
+
+final RegExp _maybeDeclarationPattern = RegExp(
+    r'\.over_react\.g\.dart'
+            // Legacy components (annotation required)
+            r'|@(?:' +
+        overReactBoilerplateAnnotations.join('|') +
+        r')'
+            // New components (references to props base classes required);
+            // find other base classes for validation purposes
+            r'|Ui(?:Factory|Props|Component|State)',
+    caseSensitive: true);
+
+/// Detects if [source] contains any key words related to boilerplate declarations.
+bool mightContainDeclarations(String source) {
+  return _maybeDeclarationPattern.hasMatch(source);
+}
+
+/// Associates boilerplate [members] with each other, grouping them into declarations and returning
+/// only the ones that require code generation.
+///
+/// If some members cannot be grouped into a valid declaration, appropriate errors will be passed to
+/// [errorCollector].
+///
+/// The order of evaluation is as follows:
+///   1. [LegacyAbstractPropsDeclaration]
+///   1. [LegacyAbstractStateDeclaration]
+///   1. [PropsMixinDeclaration]
+///   1. [StateMixinDeclaration]
+///   // None of the above include any component boilerplate entities, only props or state
+///   1. [LegacyClassComponentDeclaration] (special case for when there is only 1 of each entity in the file)
+///   1. Any of: [LegacyClassComponentDeclaration], [ClassComponentDeclaration], [PropsMapViewOrFunctionComponentDeclaration]
```

Comment:
Are all these meant to be numbered `1.`?

Reply:
:yee:, using `1.` is a way to make ordered list items in rendered markdown without having to keep the numbers up to date: https://www.markdownguide.org/basic-syntax/#ordered-lists

Example:
```markdown
1. foo
1. bar
1. baz
```
...renders as:
> 1. foo
> 1. bar
> 1. baz

Reply:
Totes, I didn't know these doc comments got rendered as markdown